### ISSUE TITLE
Allow different AMR flagging on each level and flagregion changes

### ISSUE THREAD

**Issue by rjleveque** [clawpack/geoclaw#454]
The changes I proposed in https://github.com/clawpack/amrclaw/issues/262 would naturally be carried over to GeoClaw, allowing different flagging methods and/or tolerances at each AMR level.

In addition, I propose the following:

Currently in setrun.py one can specify [flagregions](http://www.clawpack.org/dev/flagregions.html)  (either as traditional  [rectangular "regions"](http://www.clawpack.org/dev/refinement.html#specifying-amr-regions) or as [Ruled Rectangles](http://www.clawpack.org/dev/ruled_rectangles.html)).  But regions are also automatically created  to go along with every topo or dtopo file, since [the way a topofile is specified](http://www.clawpack.org/dev/setrun_geoclaw.html#topography-data-file-parameters), for example includes parameters `minlevel, maxlevel, t1, t2`.  These are used to define a rectangular region whose spatial extent is exactly that of the topofile.  For each dtopo file a region is created with spatial extent *and time interval* determined by the dtopo data and only `minlevel, maxlevel` are specified along with the path to the dtopo file.  This was originally intended to make it convenient in cases where the user wants to define a certain level of refinement around the earthquake source region, for example, over the duration of the earthquake.  

However, this capability seems to be little used in practice and much more frequently is the cause of confusion for new users.  And even experienced users often get messed up with refinement not happening where desired because of the parameters set for a topo or dtopo file were over-riding the regions that were explicitly set.

I propose we eliminate this option, so that specifying a topo or dtopo file only specifies this input data and does not implicitly generate a region.  If the user wants a region with exactly the same spatial extent as the topo file, then they should explicitly set a flagregion with this spatial extent.

Another possibility for providing this convenience:  the new way of specifying [flagregions](http://www.clawpack.org/dev/flagregions.html) allows setting either:

    flagregion.spatial_region_type = 1  # Rectangle
    flagregion.spatial_region = [x1,x2,y1,y2]

for a rectangle, or e.g.

    flagregion.spatial_region_type = 2  # Ruled Rectangle
    flagregion.spatial_region_file = \
        os.path.abspath('RuledRectangle_Trapezoid.data')

for a Ruled rectangle.  

We could also allow something like:

      flagregion.spatial_region_type = 3  # topofile extent
      flagregion.spatial_region_file = os.path.abspath('etopo1.asc')

in which case it would read the spatial extent `[x1,x2,y1,y2]` from the topofile header and set up a rectangular region.  This would allow the convenience originally intended but in a much clearer manner than the current system.

**Comment by rjleveque** [comment]
I added another suggestion to https://github.com/clawpack/amrclaw/issues/262 that would be useful in particular for GeoClaw, namely introducing `amrdata.amr_levels_max_outside_flagregions`. 

That reminded me of another GeoClaw parameter that often causes confusion and/or unexpected behavior, the parameters `deep_depth` and `max_level_deep`.  These are often confusing to people and don't work so well anyway and I propose eliminating them.

The original idea was to try to make sure that fine grids only needed near the coast should not be introduced in deeper water when it can be avoided, because the wave speed sqrt(g*h) is larger in deeper water and so small fine grids extending farther than needed into deep water potentially require much smaller time steps than if the fine grid is only very close to shore, due to the CFL restriction.

So the idea is that if the water depth is greater than `deep_depth` then a cell would not be flagged for refinement beyond level `max_level_deep`, even if it's in a region where it would be flagged otherwise.  This is confusing because it's not clear how this works in conjunction with `maxlevel` of a flagregion, for example.  Also it doesn't always work all that well -- even if a cell isn't flagged it may get refined anyway if it is part of the buffer or if it is in a rectangular patch after clustering the flagged and buffered cells.  So there might still be cells refined with water depth much greater than `deep_depth`.

It is probably better to try to keep refinement away from deep water by specifying flagregions that hug the coast, which is now much easier to do with Ruled Rectangles, e.g. [this region that follows the continental shelf](http://www.clawpack.org/dev/ruled_rectangles.html#example-covering-the-continental-shelf), or by specifying [a flagregion that covers a set of fgmax points](http://www.clawpack.org/gallery/_static/apps/notebooks/geoclaw/MarchingFront.html#amr-flag).

**Comment by mandli** [comment]
Agreed.

**Comment by mandli** [comment]
@rjleveque Thanks for the clarification.  I was thinking we would also have something apart from KML to visualize these but I had forgotten about that capability.

**Comment by rjleveque** [comment]
Yes, it would be nice to have other plotting tools too, particularly for amrclaw simulations that have nothing to do with geosciences!
